### ISSUE TITLE
Encrypted Block Volumes

### ISSUE THREAD

**Issue by JacquesMrz** [ovh/public-cloud-roadmap#307]
As a user I want to be able to encrypt my block volumes.

**Comment by scndel** [comment]
A must which would facilitate the fulfillment of some of our requirements, as far as the scope includes volumes (block storage) for _OVH Managed **Kubernetes*_* and uses server-side encryption with **customer-provided** encryption, like available now on _OVH Object Storage_ (_S3_).

**Comment by biapar** [comment]
Can I have more info about encryption on OVH Object Storage (S3)? Thanks.

**Comment by JohannesHeld** [comment]
Customized encryption of data at rest is an important feature. Especially for potential customers from the financial sector.

**Comment by JacquesMrz** [comment]
> Can I have more info about encryption on OVH Object Storage (S3)? Thanks.

Hi @biapar , here is the guide about Encryption on S3 Object Storage in OVHcloud.

**Comment by biapar** [comment]
> > Can I have more info about encryption on OVH Object Storage (S3)? Thanks.
> 
> Hi @biapar , here is the guide about Encryption on S3 Object Storage in OVHcloud.

Where?

**Comment by igorrenquin** [comment]
Hi,
 - Is this feature including : O.S. instances drive encryption?
 - setting the encryption will be including in the terraform provider?

**Comment by abrdsmygp** [comment]
Hi,

**Comment by MrOffline77** [comment]
Push - Want to see this in K8S 👍

**Comment by scndel** [comment]
Hi @JacquesMrz , it impacts our own roadmap so it'd be also nice to know that you will NOT implement it within next months (therefore we'll unfortunately aim towards another encryption solution).

**Comment by MrOffline77** [comment]
In the Openstack world there is Cinder Barbican to provide encryption at rest (LUKS) for Cinder volumes. 
As an MKS customer, I would like to be able to use transparent encryption at rest on a PV. For example via a specific annotation on the PVC. 
We solve the issue so far by running a Ceph cluster ourselves in the cluster which runs on the PV of the MKS. We then work with transparent OSD encryption and create our own storage class. In doing so, we lose approx. 50% of the IOPS. 

As a transitional solution, it would help if MKS had the high-speed GEN2 storage available in order to have more IOPS available.

**Comment by julienkosinski** [comment]
> Hi [@JacquesMrz](https://github.com/JacquesMrz) , it impacts our own roadmap so it'd be also nice to know that you will NOT implement it within next months (therefore we'll unfortunately aim towards another encryption solution).

If I read the roadmap correctly, it looks like it should be implemented more or less between July and October 2024 :).

**Comment by biapar** [comment]
Hi Julie Which Solution Do you use?
@Ovh: Why this delay?

**Comment by julienkosinski** [comment]
@biapar Well, as of now, unencrypted volumes...

**Comment by julienkosinski** [comment]
@biapar Nice! :) In this regard, depending on your use case, you might be interested in SOPS (but I advise waiting until SOPS have a clearer release cycle). The most sensitive data I have to store are on the OVH Object Storage which has a built-in encryption option. So I think to wait on OVH encryption for the Block part.

**Comment by MrOffline77** [comment]
Feel free to reach out to me for support in setting up rook with encryption at rest. We are happy to assist you.

**Comment by julienkosinski** [comment]
Thank you! :)

**Comment by fhebuterne** [comment]
Hi @JacquesMrz,

Did you have any news on this issue please ?
Disk encryption is required for DORA law has been adopted in EU for some entities (cf . https://www.digital-operational-resilience-act.com/Article_9.html and https://duokey.com/resources/chiffrement-et-dora-quelles-sont-les-obligations-imposees-par-dora)

Sincerely,
Fabien

**Comment by martinovh** [comment]
Hi, 

There are some updates, encryption on block volumes is planned for end of Q3-25 for now. I will keep you informed with furthers updates.

Sincerely, 

Martin

**Comment by karljohns0n** [comment]
@martinovh will this include encrypted instance backup as well? This is required by most customers.

**Comment by martinovh** [comment]
> [@martinovh](https://github.com/martinovh) will this include encrypted instance backup as well? This is required by most customers.

Hello, this feature is only for Block volumes for now, there will be evolutions on instance part but we don't have ETAs to communicate yet.

**Comment by jrozanski** [comment]
> > [@martinovh](https://github.com/martinovh) will this include encrypted instance backup as well? This is required by most customers.
> 
> Hello, this feature is only for Block volumes for now, there will be evolutions on instance part but we don't have ETAs to communicate yet.

I appreciate we have to start somewhere, but sadly, in case of encryption, there is no half-measures. Lack of encrypted backups/snapshots of volumes, invalidates encryption for most commercial use-cases.

**Comment by martinovh** [comment]
> > > [@martinovh](https://github.com/martinovh) will this include encrypted instance backup as well? This is required by most customers.
> > 
> > 
> > Hello, this feature is only for Block volumes for now, there will be evolutions on instance part but we don't have ETAs to communicate yet.
> 
> I appreciate we have to start somewhere, but sadly, in case of encryption, there is no half-measures. Lack of encrypted backups/snapshots of volumes, invalidates encryption for most commercial use-cases.

Hello @jrozanski, just making sure that we are on the same level of understanding, block volumes data will be encrypted and therefore they will still be encrypted on their backup/snapshot. My answer was only mentioning the lack of encryption, for now, on instances and related primary storage part.

I also have good news, Block Volumes encryption based on OVH managed key should be release by mid June 2025.

**Comment by JohannesHeld** [comment]
Hi @martinovh,

_I also have good news, Block Volumes encryption based on OVH managed key should be release by mid June 2025._ 
Are there some more information on that topic available?
Will that be available for PVC in k8s as well? Will I be able to upgrade an existing PVC to an encrypted PVC?
What might be the impact due to the overhead?

Best, Johannes

**Comment by jrozanski** [comment]
> I also have good news, Block Volumes encryption based on OVH managed key should be release by mid June 2025.

This is amazing, thank you @martinovh

**Comment by johnandonop** [comment]
Hello @martinovh, thanks a lot for the good news. We have been wanting this a lot. 

That means in about 3 weeks we will be able to encrypt volumes directly from the OVH portal. Shouldn't  there be an official communication from OVH already about this with a fixed date?  It's a great feature addition to the OVH security portfolio. 

Question: Hopefully we will be able to encrypt existing volumes and not needs to create a new one to enable encryption right?

**Comment by tgt-ops** [comment]
Any news ? It should be out by now.

**Comment by martinovh** [comment]
Hello, 

Due to technical issues, we are delaying the feature availability through OVH control panel that should be released within next weeks. 

However, creation of new encrypted with Ovh managed key "LUKS" volumes should be available through horizon control panel and Openstack CLI in the coming days in SBG5 and RBX-A regions. I will keep you informed.

Regarding the compatibility of existing volumes with encryption, users will be able to migrate non encrypted volumes to encrypted ones through retyping feature.

**Comment by martinovh** [comment]
Hello, 

As mentioned earlier this week, you are, since yesterday, able to create encrypted volume types through Horizon control panel or Openstack CLI in RBX-A and SBG5. 
Encrypted volume types are named with "-luks" at the end, so you will find 3 new volume types : Classic-luks, High-speed-luks, High-speed-gen2-luks. 

Please note that retyping is not yet working but we are working on this to deliver it as soon as possible. 

We expect to release the related OVH control panel experience in the coming weeks and from there start the rollout of the feature in other regions.

**Comment by Vatoth** [comment]
Do you have a rollout planning for other regions ? Would be great to anticipate when we can start using encrypted volumes there.

**Comment by tgt-ops** [comment]
Hello,

As I understand documentation, encrypted volumes cannot be multi AZ volumes.
Can you confirm ?

**Comment by martinovh** [comment]
Hello everyone, 

I am glad to announce that encrypted block volume with OMK (OVHcloud managed key) is now live in creations flow through OVHcloud control panel, CLI openstack and horizon control panel in : 
- RBXA 
- All SBG regions 

About coming regions in the rollout plan, it's important to know that encryption on block volumes is dependant of the availability of our KMS in targeted regions, therefore the planing is as follow  : 
- Paris will be available in the coming days 
- BHS will be released by the end of July 
- Virginia and Oregon in the US will be released by the end of August 
- DE1 & GRA will be released by October 
- WAW and ERI are expected by February 2026 

For retyping (feature allowing to move an existing non encrypted volume to an encrypted one), the feature should be available in coming weeks through horizon and CLI openstack and by the end of September through OVH control panel. 

All volume types can be retyped to encrypted equivalent types except classic multi attach in 3AZ that does not support encryption.

**Comment by svdo** [comment]
Very nice work! For my understanding: does this mean that I can now also spin up instances where even the root partition is auks-encrypted? Or is that out of scope?

**Comment by martinovh** [comment]
> > > > [@martinovh](https://github.com/martinovh) will this include encrypted instance backup as well? This is required by most customers.
> > > 
> > > 
> > > Hello, this feature is only for Block volumes for now, there will be evolutions on instance part but we don't have ETAs to communicate yet.
> > 
> > 
> > I appreciate we have to start somewhere, but sadly, in case of encryption, there is no half-measures. Lack of encrypted backups/snapshots of volumes, invalidates encryption for most commercial use-cases.
> 
> Hello [@jrozanski](https://github.com/jrozanski), just making sure that we are on the same level of understanding, block volumes data will be encrypted and therefore they will still be encrypted on their backup/snapshot. My answer was only mentioning the lack of encryption, for now, on instances and related primary storage part.
> 
> I also have good news, Block Volumes encryption based on OVH managed key should be release by mid June 2025.

Hello @svdo, as I mentioned earlier, unfortunately, this only relates to encryption of additional block volumes and not to the root partition of instances. A related issue will be created in the compute public roadmap soon, I'll add the link as soon as it is for you to be able to follow the topic.

**Comment by martinovh** [comment]
Hello everyone,

I am glad to announce that encrypted block volume with OMK (OVHcloud managed key) is now live in creations flow through OVHcloud control panel, CLI openstack and horizon control panel in **all BHS regions** (in addition to the previously mentioned).

As an updated reminder here is the rollout plan for following regions :
 
- Virginia and Oregon in the US will be released by the end of August
- DE1 & GRA will be released by October
- SGP1, SYD1 & AP-SOUTHEAST-SYD-2 will be released by November
- WAW and ERI are expected by February 2026

**Comment by johnandonop** [comment]
Hello @martinovh 

firstly i'd like to say many thanks for the continuous updates and for making this feature happen.

I confirm that from Horizon we now have the options -LUKS to create encrypted volumes. However retyping is not yet available, and i was wondering if there is a timeline on this (either from horizon or OVH Control Panel)

Thanks you!.

**Comment by pschoen-itsc** [comment]
@martinovh is there already a fixed release date for GRA?

**Comment by martinovh** [comment]
> [@martinovh](https://github.com/martinovh) is there already a fixed release date for GRA?

Hello @pschoen-itsc, I created a dedicated issue here : https://github.com/ovh/public-cloud-roadmap/issues/951
We are now targeting Mid October for the release of GRA, I'll update the dedicated issue as soon as we have more precise informations.

**Comment by jeremy-cayrasso** [comment]
@martinovh 
Encrypted block volume creation is not working with Terraform :  

as suggested by documentation : https://help.ovhcloud.com/csm/en-public-cloud-compute-storage-additional-disk?id=kb_article_view&sysparm_article=KB0050612

resource "openstack_blockstorage_volume_v3" "terraform_blockstorage" {
  name   = "testenc" 
  size   = 10               
  region = "SBG5"     
  volume_type = "high-speed-gen2-luks"

Error: Error creating openstack_blockstorage_volume_v3: Bad request with: [POST https://volume.compute.sbg5.cloud.ovh.net/v3/xxx/volumes], error message: {"badRequest": {"code": 400, "message": "Key manager error"}}

Removing -luks from the volume type works correctly and creates an unencrypted volume.

**Comment by martinovh** [comment]
Thank you for your feedback @jeremy-cayrasso , we detected an error in the validator of our terraform provider with new luks volumes. This is being fixed, I'll keep you informed when it's released.

**Comment by Kovaatchy** [comment]
@martinovh

  Hello, I'm experiencing the same "Key manager error" issue, but in a **Kubernetes context** using OVH Managed Kubernetes with the Cinder CSI driver.

  **Error message:**
  failed to provision volume with StorageClass "csi-cinder-high-speed-gen2-luks": rpc error: code = Internal desc = CreateVolume failed with error Bad request with: [POST https://volume.compute.gra7.cloud.ovh.net/v3/xxx/volumes], error message: {"badRequest": {"code": 400, "message": "Key manager error"}}

  **StorageClass configuration:**
  ```yaml
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    name: csi-cinder-high-speed-gen2-luks
  allowVolumeExpansion: true
  parameters:
    fsType: ext4
    type: high-speed-gen2-luks
  provisioner: cinder.csi.openstack.org
  reclaimPolicy: Delete
  volumeBindingMode: WaitForFirstConsumer
  ```

  Region: GRA7

  The issue appears identical to the Terraform case reported above - the -luks encrypted volume types are failing with "Key manager error". This suggests the underlying key management service might not be properly configured across different OVH regions.

  Is there any update on when encrypted volume support will be available for OVH Public Cloud?

**Comment by martinovh** [comment]
Thank you @Kovaatchy , we are currently investigating your issue and I'll keep you informed.

**Comment by geoKas91** [comment]
Hello,
Same issue in GRA9

```
failed to provision volume  with StorageClass "csi-cinder-high-speed-gen2-luks": rpc error: code = Internal desc = 
CreateVolume failed with error Bad request with: [POST https://volume.compute.gra9.cloud.ovh.net/v3/xxx/volumes], 
error message: {"badRequest": {"code": 400, "message": "Key manager error"}}
```

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-cinder-high-speed-gen2-luks
allowVolumeExpansion: true
parameters:
  fsType: ext4
  type: high-speed-gen2-luks
provisioner: cinder.csi.openstack.org
reclaimPolicy: Retain
volumeBindingMode: Immediate
```

**Comment by rverchere** [comment]
Hi, same issue there on GRA7.

**Comment by martinovh** [comment]
Hello @rverchere  @geoKas91 @Kovaatchy ,  We are currently fixing the issue, the fix should be released by the end of next week.

**Comment by martinovh** [comment]
Hello @rverchere  @geoKas91 @Kovaatchy, the fix was delivered, you should be able to create luks volumes now.

**Comment by rverchere** [comment]
@martinovh thanks, I confirm I'm able to create a luks volume now!

FYI, when I delete a luks created PVC, I need to force deletion by deleting manually the finalizer `kubernetes.io/pvc-protection`

**Comment by nicolaslacroux** [comment]
Hello,
We are pleased to announce that **LUKS encryption** is now officially supported on **Managed Kubernetes Service** (MKS).
This feature allows you to create and manage encrypted block storage volumes using LUKS via the following CSI storage classes:

- **high-speed-luks**
- **csi-cinder-classic-luks**
- **high-speed-gen2-luks**

Currently available regions:

- **France**: RBX, SBG, GRA (GRA5, GRA7, GRA9, GRA11), EU-WEST-PAR
- **Germany**: DE1
- **Italy**: EU-SOUTH-MIL
- **Canada**: BHS5
- **United States**: US-WEST-OR-1, US-EAST-VA-1

The volumes are encrypted using OVHcloud Managed Key (OMK), ensuring the confidentiality and integrity of your data at rest. 

https://github.com/ovh/public-cloud-roadmap/issues/1051

**Comment by cedric-crouzet-penbase** [comment]
Hello,

Are there any plans to fix https://github.com/ovh/public-cloud-roadmap/issues/1130 so that backup and restore work with encrypted block volumes?

Without this feature, we cannot use it in production with Velero or Kasten.
